DOCSP-35968: Transactions docs (#2847)
jenssegers/laravel-mongodb: docs/transactions.txt

using the {+odm-long+}. Transactions let you run a sequence of write operations
that update the data only after the transaction is committed.

If the transaction fails, the PHP library that manages MongoDB operations
for {+odm-short+} ensures that MongoDB discards all the changes made within
the transaction before they become visible. This property of transactions
that ensures that all changes within a transaction are either applied or
discarded is called **atomicity**.

MongoDB performs write operations on single documents atomically. If you
need atomicity in write operations on multiple documents or data consistency
across multiple documents for your operations, run them in a multi-document
transaction.

Multi-document transactions are **ACID compliant** because MongoDB
guarantees that the data in your transaction operations remains consistent,
even if the driver encounters unexpected errors.

Learn how to perform transactions in the following sections of this guide:

- :ref:`laravel-transaction-requirements`
- :ref:`laravel-transaction-callback`
- :ref:`laravel-transaction-commit`
- :ref:`laravel-transaction-rollback`

> To learn more about transactions in MongoDB, see :manual:`Transactions </core/transactions/>`
> in the {+server-docs-name+}.

Reviewer: S: you can put this in a tip admonition
```suggestion
.. tip::

   To learn more about transactions in MongoDB, see 
   :manual:`Transactions </core/transactions/>`
   in the {+server-docs-name+}.
```

.. _laravel-transaction-requirements:

Requirements and Limitations
----------------------------

> To perform transactions in MongoDB, you must be using the following MongoDB

Reviewer: S: simplify the verb tense
```suggestion
To perform transactions in MongoDB, you must use the following MongoDB
```

  transaction must be on existing collections. In MongoDB versions 4.4 and
  later, the server automatically creates collections as necessary when
  you perform write operations in a transaction. To learn more about this
  limitation, see :manual:`Create Collections and Indexes in a Transaction </core/transactions/#create-collections-and-indexes-in-a-transaction>`
  in the {+server-docs-name+}.
- MongoDB does not support nested transactions. If you attempt to start a
  transaction within another one, the extension raises a ``RuntimeException``.
  To learn more about this limitation, see :manual:`Transactions and Sessions </core/transactions/#transactions-and-sessions>`
  in the {+server-docs-name+}.
- The {+odm-long+} does not support the database testing traits
  ``Illuminate\Foundation\Testing\DatabaseTransactions`` and ``Illuminate\Foundation\Testing\RefreshDatabase``.
  As a workaround, you can create migrations with the ``Illuminate\Foundation\Testing\DatabaseMigrations``
  trait to reset the database after each test.

.. _laravel-transaction-callback:

Run a Transaction in a Callback
-------------------------------

> This section shows how you can run a transaction as a callback.

Reviewer: S: since you use the "in a callback" phrasing in the section header:
```suggestion
This section shows how you can run a transaction in a callback.
```

Author: Great catch, thanks!

When using this method of running a transaction, all the code in the
callback method runs as a single transaction.

In the following example, the transaction consists of write operations that
transfer the funds from a bank account, represented by the ``Account`` model,
to another account:

.. literalinclude:: /includes/fundamentals/transactions/TransactionsTest.php
   :language: php
   :dedent:
   :start-after: begin transaction callback
   :end-before: end transaction callback

You can optionally pass the maximum number of times to retry a failed transaction as the second parameter as shown in the following code example:

.. code-block:: php
   :emphasize-lines: 4

   DB::transaction(function() {
           // transaction code
       },
       retries: 5,
   );

.. _laravel-transaction-commit:

Begin and Commit a Transaction
------------------------------

This section shows how to start and commit a transaction.

To use this method of starting and committing a transaction, call the

> ``DB::beginTransaction()`` to start the transaction. Then, call the

Reviewer: S: Add "method"
```suggestion
``DB::beginTransaction()`` method to start the transaction. Then, call the
```

``DB::commit()`` method to end the transaction, which applies all the updates
performed within the transaction.

In the following example, the balance from the first account is moved to the
second account, after which the first account is deleted:

.. literalinclude:: /includes/fundamentals/transactions/TransactionsTest.php
   :language: php
   :dedent:
   :emphasize-lines: 1,9
   :start-after: begin commit transaction
   :end-before: end commit transaction

.. _laravel-transaction-rollback:

Roll Back a Transaction
-----------------------

This section shows how to roll back a transaction. A rollback reverts all the
write operations performed within that transaction. This means that the
data is reverted to its state before the transaction.

> To perform the roll back, call the ``DB::rollback()`` function anytime before

Reviewer: S: I think it makes sense to use "roll back" as the verb form and "rollback" as the noun form
```suggestion
To perform the rollback, call the ``DB::rollback()`` function anytime before
```
